cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -326,10 +437,22 @@ class ScalarAggregateNode : public ExecNode, public TracedNode {
   }
 
  private:
-  Status Finish() {
-    auto scope = TraceFinish();
+  Status ResetAggregates() {
+    auto exec_ctx = plan()->query_context()->exec_context();
+    for (size_t i = 0; i < kernels_.size(); ++i) {
+      const std::vector<TypeHolder>& in_types = in_typesets_[i];
+      states_[i].resize(plan()->query_context()->max_concurrency());
+      KernelContext kernel_ctx{exec_ctx};
+      RETURN_NOT_OK(Kernel::InitAll(
+          &kernel_ctx, KernelInitArgs{kernels_[i], in_types, aggs_[i].options.get()},
+          &states_[i]));
+    }
+    return Status::OK();
+  }
+
+  Status OutputResult(bool is_last = false) {
     ExecBatch batch{{}, 1};
-    batch.values.resize(kernels_.size());
+    batch.values.resize(kernels_.size() + segment_field_ids_.size());
```

Comment:
Why do we need to include segment key sizes in  the output batch size?

Reply:
These are additional keys passed to the output. For example, if you have `time` as a segment-key and `key` as a key, then both will be passed to the output.

Reply:
I am confused here. How is the size of "kernels_" related to how many key columns we have here?

Reply:
This code is in `ScalarAggregateNode`, which does not have keys but may have segment-keys. The output batch for this class has two sets of columns, one set of kernel/aggregate columns and a second set of segment-key columns. The allocation here is for both sets of columns.

Reply:
I see